src/versions.jl

```diff
@@ -0,0 +1,299 @@
+        lo1, up1 = ranges[k].lower, ranges[k].upper
+        if isjoinable(up, lo1)
+            isless_uu(up, up1) && (up = up1)
+            continue
+        end
+        vr = VersionRange(lo, up)
+        @assert !isempty(vr)
+        ranges[k0] = vr
+        k0 += 1
+        lo, up = lo1, up1
+    end
+    vr = VersionRange(lo, up)
+    if !isempty(vr)
+        ranges[k0] = vr
+        k0 += 1
+    end
+    resize!(ranges, k0 - 1)
+    return ranges
+end
+
+###############
+# VersionSpec #
+###############
+struct VersionSpec
+    ranges::Vector{VersionRange}
+    VersionSpec(r::Vector{<:VersionRange}) = new(union!(r))
+    VersionSpec(vs::VersionSpec) = new(copy(vs.ranges))
+end
+
+VersionSpec() = VersionSpec(VersionRange())
+VersionSpec(v::VersionNumber) = VersionSpec(VersionRange(v))
+VersionSpec(r::VersionRange) = VersionSpec(VersionRange[r])
+VersionSpec(s::AbstractString) = VersionSpec(VersionRange(s))
+VersionSpec(v::AbstractVector) = VersionSpec(map(VersionRange, v))
+
+# Hot code
+function Base.in(v::VersionNumber, s::VersionSpec)
+    for r in s.ranges
+        v in r && return true
+    end
+    return false
+end
+
+Base.copy(vs::VersionSpec) = VersionSpec(vs)
+
+const empty_versionspec = VersionSpec(VersionRange[])
+# Windows console doesn't like Unicode
+const _empty_symbol = @static Sys.iswindows() ? "empty" : "∅"
+
+Base.isempty(s::VersionSpec) = all(isempty, s.ranges)
+@assert isempty(empty_versionspec)
+# Hot code, measure performance before changing
+function Base.intersect(A::VersionSpec, B::VersionSpec)
+    (isempty(A) || isempty(B)) && return copy(empty_versionspec)
+    ranges = Vector{VersionRange}(undef, length(A.ranges) * length(B.ranges))
+    i = 1
+    @inbounds for a in A.ranges, b in B.ranges
+        ranges[i] = intersect(a, b)
+        i += 1
+    end
+    VersionSpec(ranges)
+end
+Base.intersect(a::VersionNumber, B::VersionSpec) = a in B ? VersionSpec(a) : empty_versionspec
+Base.intersect(A::VersionSpec, b::VersionNumber) = intersect(b, A)
+
+Base.union(A::VersionSpec, B::VersionSpec) = union!(copy(A), B)
+function Base.union!(A::VersionSpec, B::VersionSpec)
+    A == B && return A
+    append!(A.ranges, B.ranges)
+    union!(A.ranges)
+    return A
+end
+
+Base.:(==)(A::VersionSpec, B::VersionSpec) = A.ranges == B.ranges
+Base.hash(s::VersionSpec, h::UInt) = hash(s.ranges, h + (0x2fd2ca6efa023f44 % UInt))
+Base.deepcopy_internal(vs::VersionSpec, ::IdDict) = copy(vs)
+
+function Base.print(io::IO, s::VersionSpec)
+    isempty(s) && return print(io, _empty_symbol)
+    length(s.ranges) == 1 && return print(io, s.ranges[1])
+    print(io, '[')
+    for i = 1:length(s.ranges)
+        1 < i && print(io, ", ")
+        print(io, s.ranges[i])
+    end
+    print(io, ']')
+end
+Base.show(io::IO, s::VersionSpec) = print(io, "VersionSpec(\"", s, "\")")
+
+
+###################
+# Semver notation #
+###################
+
+const ver_reg = r"^([~^]?)?[v]?([0-9]+?)(?:\.([0-9]+?))?(?:\.([0-9]+?))?$"
```

Comment:
You can just write `v?` instead of `[v]?`, it's equivalent. Should this allow whitespace in some places? I guess we can start by being strict and not allowing it; we can always allow some if necessary.